polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -2192,13 +2180,16 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
                    << " ST: " << *ST << "\n";
     }
     mlir::Type ET;
-    if (auto ST = PT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>()) {
+    if (auto ST = PT.getElementType().dyn_cast<mlir::LLVM::LLVMStructType>())
       ET = ST.getBody()[fnum];
-    } else {
-      ET = PT.getElementType()
-               .cast<mlir::LLVM::LLVMArrayType>()
-               .getElementType();
-    }
+    else if (auto AT =
+                 PT.getElementType().dyn_cast<mlir::LLVM::LLVMArrayType>())
+      ET = AT.getElementType();
+    else if (auto MT = PT.getElementType().dyn_cast<MemRefType>())
+      ET = MT.getElementType();
+    else
+      llvm_unreachable("not implemented");
+
```

Comment:
Do you think a `TypeSwitch` would make this more readable? If not, leave as is.

Reply:
It would and personally I would prefer the TypeSwitch ... but I am OK with the code as is if @whitneywhtsang feels strongly about it.